### PR TITLE
Integration Tests for Retrieval models

### LINKED ISSUE #482: [FEA] Create a integration test for retrieval models based on the research experiments
# 🚀 Feature request

Create integration tests for retrieval models based on the research experiments.

## Motivation

We have created a script using advanced options of Merlin Models API to configure, build and train retrieval models (MF, Two Tower and YouTubeDNN). In addition, we performed a few hypertuning sessions to get best performance of those models for different datasets (H&M, LastFM).

This request is about creating integration tests for two reasons:
1. Fail when there are changes in our API related to Retrieval Models
2. Track over time if we have a drop in accuracy or performance (runtime) of those tests compared to the baselines (results from the research experiments).

### PR DESCRIPTION
**Pull request by gabrielspmoreira**
Fixes #482 

### Goals :soccer:
Create a set of integration tests for retrieval models to check for
- **Stability of our API** - by using advanced options of the APIs that might break if changed
- **Accuracy and Throughput** - Compute accuracy and throughput for training and evaluation, and log properly so that it is possible to track over time if we have a drop in those metrics (compared to the best hypertuned results from the retrival experiments).

### Implementation Details :construction:
- Uses [Fiddle](https://github.com/google/fiddle) for Python-first configuration management
- The retrieval training and evaluation must be available to learn in both PyTest (for CI) and also as a command line script (where hparams are provided as command line arguments)

### Testing Details :mag:
- In progress

### CODE REVIEW

**Comment by gabrielspmoreira** [comment]
rerun tests

**Comment by gabrielspmoreira** [comment]
rerun tests

**Comment by gabrielspmoreira** [comment]
rerun tests

**Comment by gabrielspmoreira** [comment]
rerun tests

File: tests/common/tf/retrieval/retrieval_utils.py [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+from merlin.models.tf.blocks.core.transformations import PopularityLogitsCorrection
+from merlin.models.utils import schema_utils
+from merlin.schema.tags import Tags
+
+
+def filter_schema(schema, filter_dataset_features=""):
+    schema = schema.select_by_tag([Tags.ITEM_ID, Tags.USER_ID, Tags.ITEM, Tags.USER])
+
+    if filter_dataset_features:
+        schema = schema.select_by_name(filter_dataset_features.split(","))
```
Comment (marcromeyn) [inline tests/common/tf/retrieval/retrieval_utils.py:40]: Now that we use Fiddle instead of a CLI, would it make sense to change this to a list instead of a string that needs to be splitted?

File: tests/common/tf/retrieval/retrieval_utils.py [hunk 2]
```diff
@@ -0,0 +1,575 @@
...
+            clipnorm=opt_clip_norm,
+            clipvalue=opt_clip_value,
+        )
+    else:
+        raise ValueError("Invalid optimizer")
+
+    return opt
+
+
+class ExamplesPerSecondCallback(tf.keras.callbacks.Callback):
```
Comment (marcromeyn) [inline tests/common/tf/retrieval/retrieval_utils.py:364]: Should we move this to merlin-models? Might be useful outside of these tests as well.

File: tests/common/tf/retrieval/retrieval_utils.py (hunk 1 above)
Comment (gabrielspmoreira) [reply tests/common/tf/retrieval/retrieval_utils.py:40]: Hey @marcromeyn . I think we can keep it as a string so that if we decide later to create a CLI based on Fiddle to run these unit tests with different arguments the string arg would be easier to parse that a list. what do you think?

File: tests/common/tf/retrieval/retrieval_utils.py (hunk 2 above)
Comment (gabrielspmoreira) [reply tests/common/tf/retrieval/retrieval_utils.py:364]: @marcromeyn  The `ExamplesPerSecondCallback` depends on `wandb` library. It is currently required only when running the integration tests. I have added the `wandb` only as dev requirement, so that CI and developers that want to check the metrics of the retrieval integration tests can do so. 
I think that maybe in the future we might having a logging module for Merlin Models, where me provide callbacks and utils for logging tools like Tensorboard, W&B, MLFlow. Then maybe then we could move this `ExamplesPerSecondCallback` to such logging module.
Does that make sense to you?

Comment (marcromeyn) [reply tests/common/tf/retrieval/retrieval_utils.py:364]: I think we can remove that dependency but we can do that in a future PR

**Review by marcromeyn — APPROVED** [review]
